### DISCUSSION TITLE
Using MuJoCo in Julia?

### DISCUSSION THREAD

**Discussion by tyler-ingebrand** [JuliaReinforcementLearning/ReinforcementLearning.jl#694] [Q&A] [answered]
Hello,

I've noticed there is a file for gym environments and MuJoCo in ReinforcementLearningEnvironments.jl/3rd_party. However, the source file gym.jl is not included in the module. When including it on my own, it throws errors at various points. 

Is MuJoCo supported by ReinforcementLearning.jl? If so, is there a special process to activate gym environments? If not, are there plans to do so in the future? 

Thanks, 
Tyler

**Comment by findmyway** [comment]
Hi Tyler,

MuJoCo used to be supported through PyCall.jl. You need to have it installed in Python first. But I haven't tested for a long time. 

Another approach is to use https://github.com/Lyceum/MuJoCo.jl directly. It's on the todo list in the next release. https://github.com/JuliaReinforcementLearning/ReinforcementLearning.jl/issues/614

**Reply by tyler-ingebrand** [reply]
Out of curiosity, is integrating Lyceum into ReinforcementLearning.jl on the todo list? Or making sure PyCall Mujoco is integrated on the todo list? Lyceum claims to use the underlying C code directly so in theory it would be faster than PyCall.

**Reply by tyler-ingebrand** [reply]
Also the errors that it is throwing are Julia compilation errors, not pycall errors.

**Reply by findmyway** [reply]
> Out of curiosity, is integrating Lyceum into ReinforcementLearning.jl on the todo list? Or making sure PyCall Mujoco is integrated on the todo list? Lyceum claims to use the underlying C code directly so in theory it would be faster than PyCall.

Lyceum



> Also the errors that it is throwing are Julia compilation errors, not pycall errors.

Could you share the error messages?

**Reply by tyler-ingebrand** [reply]
Sure, here is a MWE:

```
using ReinforcementLearning
using PyCall
include("~/.julia/packages/ReinforcementLearningEnvironments/Pzsys/src/environments/3rd_party/gym.jl") # had to do this because it is not included otherwise!

env = GymEnv("gym.envs.mujoco.humanoid_v3")
reset!(env)

for _ in 1:1000
    state = state(env)
    action = map(x -> rand(x), action_space(env))
    env(action)
    render(env)

```

Errors:

> LoadError: ArgumentError: invalid type for argument env in method definition for nameof at `/.julia/packages/ReinforcementLearningEnvironments/Pzsys/src/environments/3rd_party/gym.jl:44


If I comment out that method as I am not using it anyway, I get:

>  LoadError: ArgumentError: invalid type for argument env in method definition for copy at ~/.julia/packages/ReinforcementLearningEnvironments/Pzsys/src/environments/3rd_party/gym.jl:46

Commenting out copy as well, I get:
> LoadError: TypeError: in Type{...} expression, expected UnionAll, got a value of type typeof(GymEnv)

this error refers to the line `function (env::GymEnv{T})(action) where {T}
`

This is possibly user error, but I think something has changed with how the compiler is handling these types that it is now throwing errors. 

Thanks,
Tyler

**Reply by findmyway** [reply]
I've never seen such errors before. Could you also share which version of Julia and the version of RL.jl you are testing?

**Reply by tyler-ingebrand** [reply]
Julia 1.7.2
  [158674fc] ReinforcementLearning v0.10.0
  [de1b191a] ReinforcementLearningCore v0.8.11
  [25e41dd2] ReinforcementLearningEnvironments v0.6.12

  [438e738f] PyCall v1.93.1
  [d330b81b] PyPlot v2.10.0

**Reply by tyler-ingebrand** [reply]
Whoops! It is throwing an error because the definition of GymEnv is found in the structs file.This updated code fixes the bug, although I still get an error when trying to load an environment listed in the file:

```
using ReinforcementLearning
using PyCall
using Random
using IntervalSets
include("/home/undergrad/.julia/packages/ReinforcementLearningEnvironments/Pzsys/src/environments/3rd_party/structs.jl") # had to do this because it is not included otherwise!
include("/home/undergrad/.julia/packages/ReinforcementLearningEnvironments/Pzsys/src/environments/3rd_party/gym.jl") # had to do this because it is not included otherwise!

env = GymEnv("gym.envs.mujoco.ant_v3")
reset!(env)

for _ in 1:1000
    state = state(env)
    action = map(x -> rand(x), action_space(env))
    env(action)
    render(env)

```

```
LoadError: Gym environment gym.envs.mujoco.ant_v3 not found.

Run `ReinforcementLearningEnvironments.list_gym_env_names()` to find supported environments.

Stacktrace:
 [1] error(s::String)
   @ Base ./error.jl:33
 [2] GymEnv(name::String; seed::Nothing)
   @ Main ~/.julia/packages/ReinforcementLearningEnvironments/Pzsys/src/environments/3rd_party/gym.jl:14
 [3] GymEnv(name::String)
   @ Main ~/.julia/packages/ReinforcementLearningEnvironments/Pzsys/src/environments/3rd_party/gym.jl:4
 [4] top-level scope
   @ ~/TylerIngebrand/AlgebraicMDP/test.jl:8
 [5] include(fname::String)
   @ Base.MainInclude ./client.jl:451
 [6] top-level scope
   @ REPL[1]:1
in expression starting at /home/undergrad/TylerIngebrand/AlgebraicMDP/test.jl:8

caused by: PyError ($(Expr(:escape, :(ccall(#= /home/undergrad/.julia/packages/PyCall/7a7w0/src/pyfncall.jl:43 =# @pysym(:PyObject_Call), PyPtr, (PyPtr, PyPtr, PyPtr), o, pyargsptr, kw))))) <class 'gym.error.NameNotFound'>
NameNotFound("Environment gym.envs.mujoco.ant_v3 doesn't exist. ",)
  File "/home/undergrad/.local/lib/python3.6/site-packages/gym/envs/registration.py", line 578, in make
    _check_version_exists(ns, name, version)
  File "/home/undergrad/.local/lib/python3.6/site-packages/gym/envs/registration.py", line 209, in _check_version_exists
    _check_name_exists(ns, name)
  File "/home/undergrad/.local/lib/python3.6/site-packages/gym/envs/registration.py", line 188, in _check_name_exists
    f"Environment {name} doesn't exist{namespace_msg}. {suggestion_msg}"

Stacktrace:
  [1] pyerr_check
    @ ~/.julia/packages/PyCall/7a7w0/src/exception.jl:62 [inlined]
  [2] pyerr_check
    @ ~/.julia/packages/PyCall/7a7w0/src/exception.jl:66 [inlined]
  [3] _handle_error(msg::String)
    @ PyCall ~/.julia/packages/PyCall/7a7w0/src/exception.jl:83
  [4] macro expansion
    @ ~/.julia/packages/PyCall/7a7w0/src/exception.jl:97 [inlined]
  [5] #107
    @ ~/.julia/packages/PyCall/7a7w0/src/pyfncall.jl:43 [inlined]
  [6] disable_sigint
    @ ./c.jl:458 [inlined]
  [7] __pycall!
    @ ~/.julia/packages/PyCall/7a7w0/src/pyfncall.jl:42 [inlined]
  [8] _pycall!(ret::PyObject, o::PyObject, args::Tuple{String}, nargs::Int64, kw::Ptr{Nothing})
    @ PyCall ~/.julia/packages/PyCall/7a7w0/src/pyfncall.jl:29
  [9] _pycall!
    @ ~/.julia/packages/PyCall/7a7w0/src/pyfncall.jl:11 [inlined]
 [10] #_#114
    @ ~/.julia/packages/PyCall/7a7w0/src/pyfncall.jl:86 [inlined]
 [11] (::PyObject)(args::String)
    @ PyCall ~/.julia/packages/PyCall/7a7w0/src/pyfncall.jl:86
 [12] GymEnv(name::String; seed::Nothing)
    @ Main ~/.julia/packages/ReinforcementLearningEnvironments/Pzsys/src/environments/3rd_party/gym.jl:12
 [13] GymEnv(name::String)
    @ Main ~/.julia/packages/ReinforcementLearningEnvironments/Pzsys/src/environments/3rd_party/gym.jl:4
 [14] top-level scope
    @ ~/TylerIngebrand/AlgebraicMDP/test.jl:8
 [15] include(fname::String)
    @ Base.MainInclude ./client.jl:451
 [16] top-level scope
    @ REPL[1]:1

```

Still a bit weird I have to include the files manually. Structs is included into reinforcementLearning.jl, but structs does not include any of the other 3rd_party code, so gym is not actually usable.

**Reply by findmyway** [reply]
Other code under `3rd_party` are loaded dynamically.

https://github.com/JuliaReinforcementLearning/ReinforcementLearning.jl/blob/77b33760e7d9554a43255cd6b480345d4045bd97/src/ReinforcementLearningEnvironments/src/ReinforcementLearningEnvironments.jl#L17-L34

It's quite strange that you still need to manually include those files.

```
LoadError: Gym environment gym.envs.mujoco.ant_v3 not found.
```

This means that you haven't installed mujoco correctly through `PyCall`. You may need to make sure mujoco is installed correctly in the python environment specified in PyCall.jl

**Reply by findmyway** [reply]
I just did a quick test on the Mac with a fresh new environment of the packages you listed above, and at least I can load `CartPole-v2` successfully.

```
(test) pkg> st
      Status `~/workspace/work/test/Project.toml`
  [438e738f] PyCall v1.93.1
  [158674fc] ReinforcementLearning v0.10.0

julia> using ReinforcementLearning

julia> using PyCall

julia> env = GymEnv("CartPole-v1")
# TimeLimit

## Traits

| Trait Type        |                  Value |
|:----------------- | ----------------------:|
| NumAgentStyle     |          SingleAgent() |
| DynamicStyle      |           Sequential() |
| InformationStyle  | ImperfectInformation() |
| ChanceStyle       |           Stochastic() |
| RewardStyle       |           StepReward() |
| UtilityStyle      |           GeneralSum() |
| ActionStyle       |     MinimalActionSet() |
| StateStyle        |     Observation{Any}() |
| DefaultStateStyle |     Observation{Any}() |

## Is Environment Terminated?

No

## State Space

`Space{Vector{IntervalSets.ClosedInterval{Float32}}}(IntervalSets.ClosedInterval{Float32}[-4.8..4.8, -3.4028235e38..3.4028235e38, -0.41887903..0.41887903, -3.4028235e38..3.4028235e38])`

## Action Space

`ZeroTo(1)`

## Current State

```
Float32[0.035740428, -0.046641443, 0.022965545, -0.032434437]
```

```

**Comment by tyler-ingebrand** [comment, marked as answer]
You are correct, CartPolve-v1 works, and my MuJoCo was installed wrong. Sorry about that. 

Also, the files are actually loaded here: https://github.com/JuliaReinforcementLearning/ReinforcementLearning.jl/blob/c67a604336923ee9120d75dee74456759860fcda/src/ReinforcementLearningEnvironments/src/ReinforcementLearningEnvironments.jl

So I was wrong about that too. Thanks for the help!!!!!
